include/volume/volume_cooling_gaussians_crhmc.hpp

```diff
@@ -0,0 +1,448 @@
+    NT k = 1.0;
+    const NT tol = 0.00001;
+    bool done=false;
+    std::vector<NT> fn(N,NT(0.0));
+    std::list<Point> randPoints;
+    typedef typename std::vector<NT>::iterator viterator;
+
+    //sample N points
+    PushBackWalkPolicy push_back_policy;
+    bool raw_output = false; 
+
+    RandomPointGenerator::apply(problem, p, N, walk_length, randPoints,
+                                push_back_policy, rng, g, f, parameters, crhmc_walk, simdLen, raw_output);
+
+    while (!done)
+    {
+        NT new_a = last_a * std::pow(ratio,k);
+
+        auto fnit = fn.begin();
+        for (auto pit=randPoints.begin(); pit!=randPoints.end(); ++pit, fnit++)
+        {
+            *fnit = eval_exp(*pit, new_a)/eval_exp(*pit, last_a);
+        }
+        std::pair<NT, NT> mv = get_mean_variance(fn);
+
+        // Compute a_{i+1}
+        if (mv.second/(mv.first * mv.first)>=C || mv.first/last_ratio<1.0+tol)
+        {
+            if (k != 1.0)
+            {
+                k = k / 2;
+            }
+            done = true;
+        } else {
+            k = 2 * k;
+        }
+        last_ratio = mv.first;
+    }
+    return last_a * std::pow(ratio, k);
+}
+
+// Compute the sequence of spherical gaussians
+template
+<
+    typename WalkType,
+    typename walk_params,
+    typename RandomPointGenerator,
+    int simdLen,
+    typename Grad,
+    typename Func,
+    typename Hess,
+    typename func_params,
+    typename Input,
+    typename Polytope,
+    typename NT,
+    typename RandomNumberGenerator
+>
+void compute_annealing_schedule(Polytope& P,
+                                NT const& ratio,
+                                NT const& C,
+                                NT const& frac,
+                                unsigned int const& N,
+                                unsigned int const& walk_length,
+                                NT const& chebychev_radius,
+                                NT const& error,
+                                std::vector<NT>& a_vals,
+                                RandomNumberGenerator& rng)
+{
+
+    
+    typedef typename Polytope::PointType Point;
+    typedef typename Polytope::VT VT;
+
+    // Compute the first gaussian
+    // This uses the function from the standard volume_cooling_gaussians.hpp
+    get_first_gaussian(P, frac, chebychev_radius, error, a_vals);
+    NT a_stop = 0.0;
+    const NT tol = 0.001;
+    unsigned int it = 0;
+    unsigned int n = P.dimension();
+    const unsigned int totalSteps = ((int)150/((1.0 - frac) * error))+1;
+
+    if (a_vals[0]<a_stop) a_vals[0] = a_stop;
+
+#ifdef VOLESTI_DEBUG
+    std::cout<<"first gaussian computed "<< a_vals[0] <<std::endl;
+#endif
+
+#ifdef VOLESTI_DEBUG
+    std::cout<<"Computing the sequence of gaussians..\n"<<std::endl;
+#endif
+
+    while (true)
+    {
+
+        NT curr_fn = 0;
+        NT curr_its = 0;
+        auto steps = totalSteps;
+
+        
+        //create the crhmc problem for this variance
```

Comment:
there is a potential of optimization here, i.e. to create the problem (and the preprocessing) only once and update each time with the new variance. Is the variance needed for the preprocessing or only for sampling? 
Could we add a related TODO comment here?